packages/colorpickers/src/elements/ColorPicker/index.tsx

```diff
@@ -72,73 +70,99 @@ export interface IColorPickerProps
-    const isControlled = color !== null && color !== undefined;
-    const [uncontrolledState, dispatch] = useReducer(reducer, toState(defaultColor));
+    const [state, dispatch] = useReducer(reducer, getInitialState(color || defaultColor));
+    const previousComputedColorRef = useRef<IColor>(state.color);
+    const previousStateColorRef = useRef<IColor>(state.color);
+
+    const controlledColor = useMemo(() => {
+      if (color) {
+        if (typeof color === 'string') {
+          const convertedColor = convertStringToColor(color);
+
+          if (convertedColor) {
+            return convertedColor;
+          }
+        } else {
+          return color;
+        }
+      }
+
+      return state.color;
+    }, [color, state.color]);
+
+    useEffect(() => {
+      if (
+        !areColorsEqual(previousStateColorRef.current, state.color) &&
+        !areColorsEqual(color, state.color)
+      ) {
+        onChange && onChange(state.color);
+      }
+
+      previousStateColorRef.current = state.color;
+    }, [color, onChange, state.color]);
 
     useEffect(() => {
-      if (isControlled === false) {
-        onChange && onChange(uncontrolledState);
+      if (!areColorsEqual(controlledColor, previousComputedColorRef.current)) {
+        dispatch({ type: 'RESET_COLOR', payload: controlledColor });
       }
-    }, [uncontrolledState, onChange, isControlled]);
 
-    const state = isControlled ? toState(color) : uncontrolledState;
-    const [hexInput, setHexInput] = useState<string>(state.hex);
-    const [redInput, setRedInput] = useState<string | number>(state.red);
-    const [greenInput, setGreenInput] = useState<string | number>(state.green);
-    const [blueInput, setBlueInput] = useState<string | number>(state.blue);
-    const [alphaInput, setAlphaInput] = useState<string | number>(state.alpha);
+      previousComputedColorRef.current = controlledColor;
+    }, [controlledColor]);
 
-    const handleColorWellChange = useCallback(
-      (hsv: IHSVColor) => {
-        const hsl = hsvToHsl(hsv.h, hsv.s * 100, hsv.v * 100);
-        const rgb = hslToRgb(hsl.h, hsl.s, hsl.l);
-        const hex = rgbToString(rgb.r, rgb.g, rgb.b);
+    const handleColorWellChange = useCallback((hsv: IHSVColor) => {
+      dispatch({
+        type: 'SATURATION_CHANGE',
+        payload: hsv
+      });
+    }, []);
 
-        if (isControlled) {
-          onChange &&
-            onChange({
-              hue: hsv.h,
-              saturation: hsl.s,
-              lightness: hsl.l,
-              hex,
-              red: rgb.r,
-              green: rgb.g,
-              blue: rgb.b,
-              alpha: state.alpha
-            });
-        } else {
-          dispatch({
-            type: 'SATURATION_CHANGE',
-            payload: hsv
-          });
-        }
-      },
-      [isControlled, onChange, state.alpha]
-    );
+    const handleHueChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
```

Comment:
All color computation logic now lives within the `reducer`. For both controlled and uncontrolled usages.